### ISSUE TITLE
Override for grafana_server_addr (fact derived)

### ISSUE THREAD

**Issue by cloudguy23** [ceph/ceph-ansible#5058]
**Is your feature request related to a problem? Please describe.**
Would like to have the ability to override fact-derived value for grafana_server_addr.   Current playbook fact is derived from interface falling into the range for public_network.   

**Describe the solution you'd like**
My deployment contains public and separate cluster networks into non-routable, isolated subnets.   Each monitor, OSD, etc. are dual homed to both the public, optionally cluster networks as well as connected to a management network (this one is routed, default gw defined).   When connecting ceph-ansible deployed Grafana and connecting to deployed ceph-mgr, dashboard is unable to display Grafana dashboards since iframe is redirecting directly to grafana_server_addr:3000 -- in my case is not routable from client viewing dashboard.   

**Describe alternatives you've considered**
Manually overriding via:

`ceph dashboard set-grafana-api-url http://(custom IP):3000`

does the trick.  However has to be done every time Ceph-ansible overrides the value.

**Comment by dsavineau** [comment]
Maybe https://github.com/ceph/ceph-ansible/pull/5054 will help you even if you're not using a VIP via a loadbalancer (maybe the variable naming isn't adapted) but this will allow you to set a different IP for the grafana api url.

What do you think ?

**Comment by cloudguy23** [comment]
@dsavineau Not sure #5054 would directly address this use case.   It assumes HA proxy would be connected to the public_network (which may be fine for certain deployments).

Instead of a VIP / IP override, how about just a method to override which NIC to use as the network / IP for Grafana?  

Right now I see `roles/ceph-facts/tasksgrafana.yml` as the opportunity to implement this..  (except I don't speak Ansible as well as I'd like to):

```
- name: set grafana_server_addr fact - ipv4
  set_fact:
    grafana_server_addr: "{{ hostvars[inventory_hostname]['ansible_all_ipv4_addresses'] | ips_in_ranges(public_network.split(',')) | first }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv4'
    - dashboard_enabled | bool
    - inventory_hostname in groups[grafana_server_group_name]

- name: set grafana_server_addr fact - ipv6
  set_fact:
    grafana_server_addr: "{{ hostvars[inventory_hostname]['ansible_all_ipv6_addresses'] | ips_in_ranges(public_network.split(',')) | last | ipwrap }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv6'
    - dashboard_enabled | bool
    - inventory_hostname in groups[grafana_server_group_name]

- name: set grafana_server_addr fact - ipv4 - (external instance)
  set_fact:
    grafana_server_addr: "{{ hostvars[groups.get(grafana_server_group_name, []) | first]['ansible_all_ipv4_addresses'] | ips_in_ranges(public_network.split(',')) | first }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv4'
    - dashboard_enabled | bool
    - inventory_hostname not in groups[grafana_server_group_name]

- name: set grafana_server_addr fact - ipv6 - (external instance)
  set_fact:
    grafana_server_addr: "{{ hostvars[groups.get(grafana_server_group_name, []) | first]['ansible_all_ipv6_addresses'] | ips_in_ranges(public_network.split(',')) | last | ipwrap }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv6'
    - dashboard_enabled | bool
    - inventory_hostname not in groups[grafana_server_group_name]
```


To summarize -- Similar to how it's possible to specify interfaces for monitors in the inventory file:

e.g.

```
[mons]
mon01.localdomain.tld monitor_interface=eth1
...

[grafana-server]
grafanahost.localdomain.tld grafana_interface=eth0
```

**Comment by fmount** [comment]
Hi @cloudguy23 
> @dsavineau Not sure #5054 would directly address this use case. It assumes HA proxy would be connected to the public_network (which may be fine for certain deployments).
>
Even though the variable name is `dashboard_frontend_vip`, you can easily reuse it to override the grafana-api-url, which is a key used by the mgr to access an external grafana instance.
I'm trying to say that you can customize it with an arbitrary ip address, the user should double check it exists, it's related to the grafana deployment (e.g. haproxy use case), and it's reacheable by the mgr(s). 
> Instead of a VIP / IP override, how about just a method to override which NIC to use as the network / IP for Grafana?
> 
Not totally sure about adding this kind of fact here (like we've done for the monitors), we should discuss about this
> Right now I see `roles/ceph-facts/tasksgrafana.yml` as the opportunity to implement this.. (except I don't speak Ansible as well as I'd like to):
> 
> ```
> - name: set grafana_server_addr fact - ipv4
>   set_fact:
>     grafana_server_addr: "{{ hostvars[inventory_hostname]['ansible_all_ipv4_addresses'] | ips_in_ranges(public_network.split(',')) | first }}"
>   when:
>     - groups.get(grafana_server_group_name, []) | length > 0
>     - ip_version == 'ipv4'
>     - dashboard_enabled | bool
>     - inventory_hostname in groups[grafana_server_group_name]
> 
> - name: set grafana_server_addr fact - ipv6
>   set_fact:
>     grafana_server_addr: "{{ hostvars[inventory_hostname]['ansible_all_ipv6_addresses'] | ips_in_ranges(public_network.split(',')) | last | ipwrap }}"
>   when:
>     - groups.get(grafana_server_group_name, []) | length > 0
>     - ip_version == 'ipv6'
>     - dashboard_enabled | bool
>     - inventory_hostname in groups[grafana_server_group_name]
> 
> - name: set grafana_server_addr fact - ipv4 - (external instance)
>   set_fact:
>     grafana_server_addr: "{{ hostvars[groups.get(grafana_server_group_name, []) | first]['ansible_all_ipv4_addresses'] | ips_in_ranges(public_network.split(',')) | first }}"
>   when:
>     - groups.get(grafana_server_group_name, []) | length > 0
>     - ip_version == 'ipv4'
>     - dashboard_enabled | bool
>     - inventory_hostname not in groups[grafana_server_group_name]
> 
> - name: set grafana_server_addr fact - ipv6 - (external instance)
>   set_fact:
>     grafana_server_addr: "{{ hostvars[groups.get(grafana_server_group_name, []) | first]['ansible_all_ipv6_addresses'] | ips_in_ranges(public_network.split(',')) | last | ipwrap }}"
>   when:
>     - groups.get(grafana_server_group_name, []) | length > 0
>     - ip_version == 'ipv6'
>     - dashboard_enabled | bool
>     - inventory_hostname not in groups[grafana_server_group_name]
> ```
> 
> To summarize -- Similar to how it's possible to specify interfaces for monitors in the inventory file:
> 
> e.g.
> 
> ```
> [mons]
> mon01.localdomain.tld monitor_interface=eth1
> ...
> 
> [grafana-server]
> grafanahost.localdomain.tld grafana_interface=eth0
> ```

**Comment by cloudguy23** [comment]
What about something simpler..   Can't we just override the `grafana_server_addr` in all.yml?

I got the following to work from roles/ceph-facts/tasks/grafana.yml:

```
- name: set grafana_server_addr fact - ipv4
  set_fact:
    grafana_server_addr: "{{ hostvars[inventory_hostname]['ansible_all_ipv4_addresses'] | ips_in_ranges(public_network.split(',')) | first }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv4'
    - grafana_server_addr is undefined
    - dashboard_enabled | bool
    - inventory_hostname in groups[grafana_server_group_name]


- name: set grafana_server_addr fact - ipv6
  set_fact:
    grafana_server_addr: "{{ hostvars[inventory_hostname]['ansible_all_ipv6_addresses'] | ips_in_ranges(public_network.split(',')) | last | ipwrap }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv6'
    - grafana_server_addr is undefined
    - dashboard_enabled | bool
    - inventory_hostname in groups[grafana_server_group_name]

- name: set grafana_server_addr fact - ipv4 - (external instance)
  set_fact:
    grafana_server_addr: "{{ hostvars[groups.get(grafana_server_group_name, []) | first]['ansible_all_ipv4_addresses'] | ips_in_ranges(public_network.split(',')) | first }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv4'
    - dashboard_enabled | bool
    - inventory_hostname not in groups[grafana_server_group_name]

- name: set grafana_server_addr fact - ipv6 - (external instance)
  set_fact:
    grafana_server_addr: "{{ hostvars[groups.get(grafana_server_group_name, []) | first]['ansible_all_ipv6_addresses'] | ips_in_ranges(public_network.split(',')) | last | ipwrap }}"
  when:
    - groups.get(grafana_server_group_name, []) | length > 0
    - ip_version == 'ipv6'
    - dashboard_enabled | bool
    - inventory_hostname not in groups[grafana_server_group_name]
```

The above properly configures the container IP, however still have to change the dashboard grafana IP manually post ceph-ansible:

`ceph dashboard set-grafana-api-url http://NEW_IP:3000`

**Comment by Hobadee** [comment]
> What about something simpler.. Can't we just override the grafana_server_addr in all.yml?
Seconded.

Since we are using Ceph purely for VMware shared storage, we have segregated our Ceph networks completely; they aren't even privately route-able.  We have an additional management interface which is firewalled off but gives authorized users access to the Ceph management GUI and the machines.  This setup works great except for Grafana, which binds to the unroutable Ceph subnet.

Passing grafana_server_addr as an Ansible `-x` parameter fixes this, but is annoying.

**Comment by dsavineau** [comment]
> but gives authorized users access to the Ceph management GUI and the machines. This setup works great except for Grafana, which binds to the unroutable Ceph subnet.

Why do you need to access to Grafana if you already have access to te Ceph Dashboard (which embed the Grafana dashboards) ?

**Comment by cloudguy23** [comment]
This would be simpler is the ceph-mgr dashboard proxied grafana requests vs. referencing them.   In my particular setup public is an isolated /protected network anyway, so no issues with unauthorized (ie public) grafana access.  ideally, grafana should be authenticated, and proxied.   However I gather that's not the scope of ceph-ansible.   

@dsavineau a simple override as suggested above (`grafana_server_addr`)as part of the playbooks would be more than sufficient.

**Comment by Hobadee** [comment]
> Why do you need to access to Grafana if you already have access to te Ceph Dashboard (which embed the Grafana dashboards) ?

The embedded dashboards are just iframes to the Grafana dashboards.  The browser just hangs while attempting to connect to $PRIVATE_IP.
